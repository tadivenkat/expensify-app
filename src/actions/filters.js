
export const setTextFilterAction = (text = '') => {
    return {
        type: 'SET_TEXT_FILTER',
        payload: text
    }
};

export const setSortByFilterAction = (sortBy = '') => {
    return {
        type: 'SET_SORT_BY',
        payload: sortBy
    };
}

export const setStartDateFilterAction = (startDate) => {
    return {
        type: 'SET_START_DATE',
        payload: startDate
    };
}

export const setEndDateFilterAction = (endDate) => {
    return {
        type: 'SET_END_DATE',
        payload: endDate
    }
}