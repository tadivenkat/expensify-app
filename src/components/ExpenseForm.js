import React from 'react';
import moment from 'moment';
import DatePicker from 'react-datepicker';


class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expense ? props.expense.id : '',
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount / 100 : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt).valueOf() : moment().valueOf(),
            error: ''
        };
    };
    
    onDescriptionChange = (event) => {
        const description = event.target.value;
        this.setState((currentState) => {
            return {
                description
            };
        });
    };

    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState((currentState) => {
            return {
                note
            };
        });
    };

    onAmountChange = (event) => {
        const amount = event.target.value;
        if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState((currentState) => {
                return {
                    amount
                };
            });
        }
    }

    onDateChange = (date) => {
        this.setState((currentState) => {
            return {
                createdAt: moment(date).valueOf()
            };
        });
    }

    onFormSubmit = (event) => {
        event.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => {
               return {
                   error: 'Please enter description and amount' 
               };
            });
        } else {
            this.setState(() => {
                return {
                    error: '' 
                };
            });
            // Construct expense object and set it on props to make it accessible to the parent component
            this.props.onSubmit({
                id: this.state.id,
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt
            });
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onFormSubmit}>
                {this.state.error && <p className="bg-danger text-white">{this.state.error}</p>}
                <div className="form-group">
                    <label for="description">Description</label>
                    <input 
                        type="text"
                        id="description"
                        className="form-control"
                        placeholder="Enter Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange} />
                </div>
                <div className="form-group">
                    <label for="amount">Amount</label>
                    <input
                        id="amount"
                        type="number"
                        className="form-control" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}/>
                </div>
                <div className="form-group">
                    <label for="note">Note</label>
                    <textarea 
                        id="note"
                        name="note"
                        className="form-control" 
                        placeholder="Add a Note for this Expense"
                        rows="7"
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                </div>
                <div className="form-group">
                    <label for="createdAt">Created At</label>
                    <DatePicker
                        id="createdAt"
                        name="createdAt"
                        className="form-control" 
                        todayButton={"Today"}
                        selected={new Date(this.state.createdAt)} 
                        onChange={this.onDateChange}
                        title="Created At"
                        dateFormat="dd-MMM-yyyy"/>
                </div>
                <div>
                    <button onClick={this.onFormSubmit} className="btn btn-primary">Save Expense</button>
                </div>                    
            </form>                                        
        );
    }
}

export default ExpenseForm;