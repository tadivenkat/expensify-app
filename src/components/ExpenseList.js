import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import filterExpenses from '../selectors/expenses';
import {totalExpenditure} from '../selectors/expenses';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseList = (props) => {
    const {text, sortBy, startDate, endDate} = props.filters;
    return (
        <div>
            <h1>Expense List</h1>
            <p>
                You have {props.expenses.length} {props.expenses.length === 1 ? 'expense' : 'expenses'} with a total 
                expenditure of {numeral(totalExpenditure(props.expenses)/100).format('$ 0,0.00')}
            </p>
            {props.expenses.map((expense) => <ExpenseListItem 
                                                    key={expense.id} 
                                                    expense={expense}/>)}
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: filterExpenses(state.expenses, state.filters),
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseList);