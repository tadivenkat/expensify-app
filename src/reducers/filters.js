
const filtersDefaultState = {
    text: '',
    sortBy: 'date',
    startDate: undefined,
    endDate: undefined
}

export default (state = filtersDefaultState, action) => {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.payload
            };
        case 'SET_SORT_BY':
            return {
                ...state,
                sortBy: action.payload
            };
        case 'SET_START_DATE':
            return {
                ...state,
                startDate: action.payload
            };
        case 'SET_END_DATE':
            return {
                ...state,
                endDate: action.payload
            };
        default: 
            return state;
    }
};