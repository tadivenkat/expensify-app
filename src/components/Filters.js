import React from 'react';
import {connect} from 'react-redux';
import {setTextFilterAction, setStartDateFilterAction, setEndDateFilterAction} from '../actions/filters';
import {setSortByFilterAction} from '../actions/filters';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import moment from 'moment';


export class Filters extends React.Component {
    setTextFilterAction = (event) => {
        this.props.setTextFilterAction(event.target.value);
    }
    setStartDateFilterAction = (date) => {
        this.props.setStartDateFilterAction(date);
    }
    setEndDateFilterAction = (date) => {
        this.props.setEndDateFilterAction(date);
    }
    setSortByFilterAction = (event) => {
        this.props.setSortByFilterAction(event.target.value);
    }
    onSubmit = (event) => {
        event.preventDefault();
    }
    render() {
        const filters = this.props.filters;
        return (
            <form class="form-inline" onSubmit={this.onSubmit}>
                <input 
                    type="text"
                    class="form-control"
                    name="TextFilter" 
                    placeholder="Search Expenses"
                    value={filters.text} 
                    onChange={this.setTextFilterAction}/>
                <select class="form-control" value={filters.sortBy} onChange={this.setSortByFilterAction}>
                    <option value="amount">Amount</option>
                    <option value="date">Date</option>
                </select>
                <DatePicker
                    class="form-control"
                    todayButton={"Today"}
                    selectsStart 
                    selected={moment(filters.startDate).valueOf()}
                    startDate={moment(filters.startDate).valueOf()}
                    endDate={moment(filters.endDate).valueOf()} 
                    onChange={this.setStartDateFilterAction}
                    title="Start Date"
                    dateFormat="dd-MMM-yyyy"/>
                <DatePicker
                    class="form-control"
                    todayButton={"Today"}
                    selectsEnd
                    selected={moment(filters.endDate).valueOf()} 
                    onChange={this.setEndDateFilterAction}
                    title="End Date"
                    dateFormat="dd-MMM-yyyy"/>
            </form>
        );
    }
};

const mapStateToProps = (state, props) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        setTextFilterAction: (text) => dispatch(setTextFilterAction(text)),
        setStartDateFilterAction: (date) => dispatch(setStartDateFilterAction(moment(date).valueOf())),
        setEndDateFilterAction: (date) => dispatch(setEndDateFilterAction(moment(date).valueOf())),
        setSortByFilterAction: (sortBy) => dispatch(setSortByFilterAction(sortBy))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Filters);