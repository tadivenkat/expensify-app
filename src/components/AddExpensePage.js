import React from 'react';
import {connect} from 'react-redux';
import {addExpenseAction} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/');            
    }    
    render() {
        return (
            <div>
                <h1>Add Expense Page</h1>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(addExpenseAction(expense))
    };
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);