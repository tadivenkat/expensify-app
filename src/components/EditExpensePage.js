import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpenseAction, removeExpenseAction} from '../actions/expenses';

const EditExpensePage = (props) => {
    return (
        <div>
            <h1>Editing Expense: {props.match.params.id}</h1>
            <ExpenseForm 
                expense={props.expense}
                onSubmit={(expense) => {
                    console.log(expense);
                    props.dispatch(editExpenseAction(expense));
                    props.history.push('/');
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpenseAction(props.expense));
                props.history.push("/");
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps)(EditExpensePage);