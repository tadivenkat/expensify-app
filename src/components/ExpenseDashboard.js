import React from 'react';
import ExpenseList from './ExpenseList';
import Filters from './Filters';
import Header from './Header';

class ExpenseDashboard extends React.Component {
  render() {
    return (
      <div>
        <Header/>
        <Filters/>
        <ExpenseList/>
      </div>
    );
  }
}

export default ExpenseDashboard;
