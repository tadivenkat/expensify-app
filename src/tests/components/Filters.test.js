import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import {Filters} from '../../components/Filters';
import filters from '../fixtures/filters';

let setTextFilterAction, setStartDateFilterAction, setEndDateFilterAction, setSortByFilterAction, wrapper;
beforeEach(() => {
    setTextFilterAction = jest.fn();
    setStartDateFilterAction = jest.fn();
    setEndDateFilterAction = jest.fn();
    setSortByFilterAction = jest.fn();
    wrapper = shallow(<Filters 
                        setTextFilterAction={setTextFilterAction}
                        setStartDateFilterAction={setStartDateFilterAction}
                        setEndDateFilterAction={setEndDateFilterAction}
                        setSortByFilterAction={setSortByFilterAction}
                        filters={filters}/>);
});

test('Should render Filters component correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should handle text change', () => {
    const value = 'Marathon';
    wrapper.find('input').simulate('change', {target: {value}});
    expect(setTextFilterAction).toHaveBeenLastCalledWith(value);
});

test('Should sort by date', () => {
    const value = 'date';
    wrapper.find('select').simulate('change', {target: {value}});
    expect(setSortByFilterAction).toHaveBeenLastCalledWith(value);
});

test('Should sort by amount', () => {
    const value = 'amount';
    wrapper.find('select').simulate('change', {target: {value}});
    expect(setSortByFilterAction).toHaveBeenLastCalledWith(value);
});

test('Should handle start date change', () => {
    const value = 1000;
    wrapper.find('DatePicker').at(0).simulate('change', {date: value});
    expect(setStartDateFilterAction).toHaveBeenLastCalledWith({date: value});
});

test('Should handle end date change', () => {
    const value = 100001;
    wrapper.find('DatePicker').at(1).simulate('change', {date: value});
    expect(setEndDateFilterAction).toHaveBeenLastCalledWith({date: value});
});