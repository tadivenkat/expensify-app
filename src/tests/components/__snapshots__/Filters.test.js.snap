// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Filters component correctly 1`] = `
<div>
  <input
    name="TextFilter"
    onChange={[Function]}
    placeholder="Filter by text"
    type="text"
    value=""
  />
  <DatePicker
    allowSameDay={false}
    dateFormat="dd-MMM-yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    endDate={1000000}
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={0}
    selectsStart={true}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    startDate={0}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    title="Start Date"
    todayButton="Today"
    withPortal={false}
  />
  <DatePicker
    allowSameDay={false}
    dateFormat="dd-MMM-yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={1000000}
    selectsEnd={true}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    title="End Date"
    todayButton="Today"
    withPortal={false}
  />
  <select
    onChange={[Function]}
    value="amount"
  >
    <option
      value="amount"
    >
      Amount
    </option>
    <option
      value="date"
    >
      Date
    </option>
  </select>
</div>
`;
